cloud/scope/identity.go

```diff
@@ -0,0 +1,162 @@
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package scope
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/Azure/go-autorest/autorest"
+	"github.com/Azure/go-autorest/autorest/adal"
+	"github.com/pkg/errors"
+	infrav1 "sigs.k8s.io/cluster-api-provider-azure/api/v1alpha3"
+	"sigs.k8s.io/cluster-api-provider-azure/util/identity"
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1alpha3"
+	clusterctl "sigs.k8s.io/cluster-api/cmd/clusterctl/api/v1alpha3"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	aadpodv1 "github.com/Azure/aad-pod-identity/pkg/apis/aadpodidentity/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// AzureCredentialsProvider provides
+type AzureCredentialsProvider struct {
+	Client       client.Client
+	AzureCluster *infrav1.AzureCluster
+	Identity     *infrav1.AzureClusterIdentity
+}
+
+// NewAzureCredentialsProvider creates a new AzureCredentialsProvider from the supplied inputs.
+func NewAzureCredentialsProvider(ctx context.Context, kubeClient client.Client, azureCluster *infrav1.AzureCluster) (*AzureCredentialsProvider, error) {
+	if azureCluster.Spec.IdentityRef == nil {
+		return nil, errors.New("failed to generate new AzureCredentialsProvider from empty identityName")
+	}
+
+	ref := azureCluster.Spec.IdentityRef
+	// if the namespace isn't specified then assume it's in the same namespace as the AzureCluster
+	namespace := ref.Namespace
+	if namespace == "" {
+		namespace = azureCluster.Namespace
+	}
+	identity := &infrav1.AzureClusterIdentity{}
+	key := client.ObjectKey{Name: ref.Name, Namespace: namespace}
+	if err := kubeClient.Get(ctx, key, identity); err != nil {
+		return nil, errors.Errorf("failed to retrieve AzureClusterIdentity external object %q/%q: %v", key.Namespace, key.Name, err)
+	}
+
+	if identity.Spec.Type != infrav1.ServicePrincipal {
+		return nil, errors.New("AzureClusterIdentity is not of type Service Principal")
+	}
+
+	return &AzureCredentialsProvider{
+		Client:       kubeClient,
+		AzureCluster: azureCluster,
+		Identity:     identity,
+	}, nil
+}
+
+// GetAuthorizer returns Azure authorizer based on the provided azure identity
+func (p *AzureCredentialsProvider) GetAuthorizer(ctx context.Context, resourceManagerEndpoint string) (autorest.Authorizer, error) {
+	azureIdentityType, err := getAzureIdentityType(p.Identity)
+	if err != nil {
+		return nil, err
+	}
+	copiedIdentity := &aadpodv1.AzureIdentity{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       "AzureIdentity",
+			APIVersion: "aadpodidentity.k8s.io/v1",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      identity.GetAzureIdentityName(p.AzureCluster.Name, p.AzureCluster.Namespace, p.Identity.Name),
```

Comment:
if this has the cluster name in it, its lifecycle is owned by the life of the cluster, no? Why not create/delete it when the an AzureCluster is created/deleted, similarly to how we create/delete the azure.json secret?

Reply:
I just added the cluster name to make it easier to find, but it wasn't there before, its not part of the cluster lifecycle it lives in a different namespace and is for internal use by the controller